prow/gcsupload/run.go

```diff
@@ -38,21 +38,35 @@ import (
 // to their destination in GCS, so the caller can
 // operate relative to the base of the GCS dir.
 func (o Options) Run(extra map[string]gcs.UploadFunc) error {
-	var builder gcs.RepoPathBuilder
-	switch o.PathStrategy {
-	case kube.PathStrategyExplicit:
-		builder = gcs.NewExplicitRepoPathBuilder()
-	case kube.PathStrategyLegacy:
-		builder = gcs.NewLegacyRepoPathBuilder(o.DefaultOrg, o.DefaultRepo)
-	case kube.PathStrategySingle:
-		builder = gcs.NewSingleDefaultRepoPathBuilder(o.DefaultOrg, o.DefaultRepo)
-	}
-
 	spec, err := downwardapi.ResolveSpecFromEnv() // TODO: pass in all the config instead of needing this?
 	if err != nil {
 		return fmt.Errorf("could not resolve job spec: %v", err)
 	}
 
+	uploadTargets := o.assembleTargets(spec, extra)
+
+	if !o.DryRun {
+		ctx := context.Background()
+		gcsClient, err := storage.NewClient(ctx, option.WithCredentialsFile(o.GcsCredentialsFile))
+		if err != nil {
+			return fmt.Errorf("could not connect to GCS: %v", err)
+		}
+
+		if err := gcs.Upload(gcsClient.Bucket(o.Bucket), uploadTargets); err != nil {
```

Comment:
Where's the issue for allowing any gcs prefix for a root, not just a bucket? This will make me cringe every time I read it, and make me frustrated with the tools arbitrary restrictions.

Also, wasn't there a PR to fix this?

Can we put a `TODO(#1234): foo` type note at least?

Reply:
This PR simply adds tests. A separate PR went in to expose to USERS the nice GCS path thing -- we need the bucket name here, literally, so we decompose the URL they give into the bits we need. Not clear what is wrong ere.